### PR TITLE
escape.d: deduplicate arg/param iteration logic

### PR DESCRIPTION
**Pull request by dkorpel**
Reduces indentation, so best reviewed with whitespace off

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by dkorpel** [comment]
> core.exception.ArrayIndexError@D:\a\1\s\compiler\src\dmd\root\array.d(131): index [172292160] is out of bounds for array of length 94532592

Yikes

File: compiler/src/dmd/escape.d [hunk 1]
```diff
@@ -2171,6 +2144,29 @@ void escapeByRef(Expression e, EscapeByResults* er, bool live = false, bool retR
...
+/**
+ * Iterate over the arguments and corresponding parameters,
+ * ignoring the hidden `_arguments[]` argument of a DstyleVariadic.
+ *
+ * Params:
+ *   tf = function type
+ *   args = arguments
+ *   dg = delegate to call for each argument
+ */
+private void foreachEscapableArgument(TypeFunction tf, Expressions* args, void delegate(Parameter, Expression) dg) {
```
Comment (WalterBright) [inline compiler/src/dmd/escape.d:2156]: Use the usual bracing style.

File: compiler/src/dmd/escape.d [hunk 2]
```diff
@@ -2171,6 +2144,29 @@ void escapeByRef(Expression e, EscapeByResults* er, bool live = false, bool retR
...
+    if (!args)
+        return;
```
Comment (WalterBright) [inline compiler/src/dmd/escape.d:2158]: Move away from passing args by pointer to by ref. This means the above test will have to be done before the function is called.

File: compiler/src/dmd/escape.d [hunk 3]
```diff
@@ -2171,6 +2144,29 @@ void escapeByRef(Expression e, EscapeByResults* er, bool live = false, bool retR
...
+
+    // j=1 if _arguments[] is first argument,
+    // skip it because it is not passed by ref
+    int j = tf.isDstyleVariadic();
```
Comment (WalterBright) [inline compiler/src/dmd/escape.d:2162]: const size_t

File: compiler/src/dmd/escape.d [hunk 4]
```diff
@@ -2171,6 +2144,29 @@ void escapeByRef(Expression e, EscapeByResults* er, bool live = false, bool retR
...
+    for (size_t i = j; i < args.length; ++i)
+    {
+        if (i - j < tf.parameterList.length && i >= j)
```
Comment (WalterBright) [inline compiler/src/dmd/escape.d:2165]: ParameterList.length is a complex function, factor it out of the loop

Comment (WalterBright) [reply compiler/src/dmd/escape.d:2165]: i >= j is always true

Comment (WalterBright) [reply compiler/src/dmd/escape.d:2165]: put in an assert to make sure!

**Comment by RazvanN7** [comment]
@dkorpel rebase this and let's get it in.

**Comment by dkorpel** [comment]
The failure on OMF is persistent

**Comment by dkorpel** [comment]
> The failure on OMF is persistent

Well guess what's gone now :slightly_smiling_face: 

But it has merge conflicts and want to refactor this a different way now, so I'll open a new PR for that.

**Pull request closed without merging**